File: GoL/Model/Game.m
//
//  Game.m
//  GoL
//
//  Created by Georg Meyer on 27.04.14.
//
//

#import "Cell.h"
#import "Game.h"
#import "GameInitializer.h"
#import "Dimensions.h"

#import "GameStateInitializer.h"

@interface Game ()

@property (strong, nonatomic) NSArray *field;
@property (strong, nonatomic) Dimensions *dimensions;

@property (strong, nonatomic) GameStateInitializer *gameStateInitializer;

@end

@implementation Game

- (id)initWithDimensions:(Dimensions *)dimensions andGameStateInitializer:(GameStateInitializer *)gameStateInitializer {
    self = [super init];
    if (self) {
        self.dimensions = dimensions;
        self.gameStateInitializer = gameStateInitializer;
        [self createField];
    }
    return self;
}

- (void)createField {
    self.field = [[[GameInitializer alloc] initWithInitialState:[self.gameStateInitializer generateGameState] andDimensions:[self dimensions]] initialize];
}

- (void)tick {
    [self eachCell: ^(Cell *cell) {
        [cell storeNextState];
    }];
    [self eachCell: ^(Cell *cell) {
        [cell applyNextState];
    }];
}

- (void)addFigure:(Figure *)figure {
    int y = 10;
    for(int i = 0; i < figure.dimensions.height; i++) {
        int x = 10;
        for(int j = 0; j < figure.dimensions.width; j++) {
            Cell *cell = [[self.field objectAtIndex:y] objectAtIndex:x];
            NSLog(@"cell");
            if([[[figure.field objectAtIndex:i] objectAtIndex:j] boolValue]) {
                [cell becomeAlive];
            } else{
                [cell die];
            }
            x++;
        }
        y++;
    }
}

- (void)eachCell:(void (^)(Cell *cell))action {
    for (NSArray *row in self.field) {
        for (Cell *cell in row) {
            action(cell);
        }
    }
}

- (NSString *)description {
    NSMutableString *result = [[NSMutableString alloc] init];
    for (NSArray *row in self.field) {
        for (Cell *cell in row) {
            [result appendString:[cell description]];
        }

        [result appendString:@"\n"];
    }
    return result;
}

@end

File: GoL/Model/Game.swift
//
//  Game.swift
//  GoL
//
//  Created by Georg Meyer on 08.06.14.
//
//

import Foundation

class Game {
    
    var dimensions : Dimensions
    var field : Array<Array<Cell>>
    var gameStateInitializer : GameStateInitializer
    var description : String{
        var result = ""
        
        for row in field {
            for cell in row {
                result += cell.description
            }
            result += "\n"
        }
            
        return result
    }
    
    
    init(dimensions : Dimensions, gameStateInitializer : GameStateInitializer){
        self.dimensions = dimensions
        self.gameStateInitializer = gameStateInitializer
        self.field = GameInitializer(initialState: gameStateInitializer.generateGameState(), dimensions: dimensions).initialize()
    }
    
    func tick() {
        for row in field {
            for cell in row {
                cell.storeNextState()
            }
        }
        for row in field {
            for cell in row {
                cell.applyNextState()
            }
        }
    }
    
    func addFigure(figure : Figure){
        
    }
    
    
}